test/4_test.py

```diff
@@ -0,0 +1,315 @@
+import logging
+import requests
+import uuid
+import urllib
+import sys
+import time
+
+class Test:
+    def __init__(self):
+        self.access_token = None
+        self.room_id = None
+
+    def _item_generator(self, json_input, lookup_key):
+        if isinstance(json_input, dict):
+            for k, v in json_input.items():
+                if k == lookup_key:
+                    yield v
+                else:
+                    for child_val in self._item_generator(v, lookup_key):
+                        yield child_val
+        elif isinstance(json_input, list):
+            for item in json_input:
+                for item_val in self._item_generator(item, lookup_key):
+                    yield item_val
+
+
+    # Further setup
+    def setup(self):
+        print('Authenticating')
+
+        authentication = requests.post('http://localhost:8080/_matrix/client/r0/login', json = {
+            'type': 'm.login.password',
+            'identifier': {
+                'type': 'm.id.user',
+                'user': 'user_1'
+            },
+            'password': 'user_1',
+        }).json()
+        print('Authentication %s' % authentication)
+        self.access_token = 'Bearer %s' % authentication['access_token']
+
+
+        # 1b. Creating the room.
+
+        print('Creating public room')
+
+        room_creation = requests.post('http://localhost:8080/_matrix/client/r0/createRoom', json = {
+            'visibility': 'public',
+        }, headers = {
+            'Authorization': self.access_token
+        }).json()
+        print('Room creation %s' % room_creation)
+        self.room_id = room_creation['room_id']
+
+        print('Created room %s' % self.room_id)
+        pass
+
+    # 2. Starting actual test
+    def test(self):
+        quoted_room_id = urllib.parse.quote("%s" % self.room_id)
+        print("Testing with room %s, access token %s" % (quoted_room_id, self.access_token))
+
+        # A set of good events. They should be preserved in history.
+        good_events = {}
+        # A set of bad events. They should either be intercepted before they reach history or be redacted from history.
+        bad_events = {}
```

Comment:
These are both dicts, not sets. I'm unsure if you want them to be sets (via `set()`) or if you want them to be dicts mapping event IDs to descriptions it seems?